pkg/resource/restconfig_factory.go

```diff
@@ -2,30 +2,115 @@ package resource
 	"k8s.io/client-go/discovery"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/restmapper"
 	"k8s.io/client-go/tools/clientcmd"
 	cmdutil "k8s.io/kubectl/pkg/cmd/util"
+	"sync"
+	"time"
 )
 
+// HIVE-2399: In the cluster sync controller, getRESTConfigFactory is called once per
+// reconcile loop. Previously, a restConfigClientGetter was created for every reconcile,
+// used to apply (or rather, attempt to apply) various synced objects, and then thrown
+// away at the end of the reconcile pass.
+//
+// The restConfigClientGetter (or more accurately, the  cmdutil.Factory that wraps it) is
+// used by the cluster sync controller (via this helper) to apply the various objects to
+// the target cluster during the syncing process. During the application process (see apply.go),
+// the apply command is set up (aptly named setupApplyCommand()) using the factory returned
+// by the getRESTConfigFactory() below. Within this setup, the ToRESTMapper() function of the
+// restConfigClientGetter is invoked. While the mapper returned is a deferred variant (restmapper.DeferredDiscoveryRESTMapper),
+// eventually the inner NewDiscoveryRESTMapper() function is invoked, which is very expensive
+// and allocates a huge amount of memory over time if repeatedly called in a hot loop. A graphic
+// demonstrating these allocations is attached the HIVE-2399 card.
+//
+// The solution is twofold:
+// 1) store the deferred mapper on the restConfigClientGetter so that it can be reused
+// 2) cache the restConfigClientGetter for each rest config and reuse it for future reconciles
+// The type of this map is effectively map[string]*restConfigClientGetter
+var restConfigClientGetterCache sync.Map
+
+var cacheCleanMu sync.Mutex
+var lastCacheClean time.Time
+var cacheCleanInterval = time.Hour * 4
+var cacheEntryMaxAge = time.Hour * 24
+
+func init() {
+	// TODO: override cache duration variables from env
+}
+
+func RemoveRestConfigClientGetterCacheEntry(logger log.FieldLogger, key string) {
+	logger.WithField("getter-cache-key", key).Debug("removing client config cache entry")
+
+	restConfigClientGetterCache.Delete(key)
+}
+
 func (r *helper) getRESTConfigFactory(namespace string) (cmdutil.Factory, error) {
 	if r.metricsEnabled {
 		// Copy the possibly shared restConfig reference and add a metrics wrapper.
 		cfg := rest.CopyConfig(r.restConfig)
 		controllerutils.AddControllerMetricsTransportWrapper(cfg, r.controllerName, false)
 		r.restConfig = cfg
 	}
-	r.logger.WithField("cache-dir", r.cacheDir).Debug("creating cmdutil.Factory from REST client config and cache directory")
-	f := cmdutil.NewFactory(&restConfigClientGetter{restConfig: r.restConfig, cacheDir: r.cacheDir, namespace: namespace})
+
+	// If we're not using the cache, this will be the getter used. If we are using the cache,
+	// this value will be stored if the cache misses.
+	freshRestClientGetter := &restConfigClientGetter{
+		restConfig: r.restConfig,
+		cacheDir:   r.cacheDir,
```

Comment:
I'm curious what this cacheDir is for. Does its function overlap with the caching we're implementing here?

[Later] Looks like it might be for schema, which would be cool (and hopefully unrelated??)

Reply:
It was there before, I just kept it there.  No idea what it does